### PR TITLE
Re-enable LOOKUP JOIN tests in 8.18

### PR DESCRIPTION
**Pull request by craigtaverner**
When we back-ported the `LOOKUP JOIN` PRs to 8.x (see https://github.com/elastic/elasticsearch/pull/117967), we found it necessary to disable all csv-spec tests since they create indices with mode:lookup, which is illegal in the cluster state of mixed clusters where other nodes do not understand the new index mode. We need to re-enable the tests, and make sure the tests are only disabled in mixed clusters with node versions too old to handle the new mode.

### CODE REVIEW

File: x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/MultiClusterSpecIT.java [hunk 1]
```diff
@@ -280,4 +280,10 @@ protected boolean enableRoundingDoubleValuesOnAsserting() {
     protected boolean supportsInferenceTestService() {
         return false;
     }
+
+    @Override
+    protected boolean supportsIndexModeLookup() throws IOException {
+        // return hasCapabilities(List.of("join_lookup_v4"));
```
Comment (craigtaverner) [inline x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/MultiClusterSpecIT.java:286]: CCS does not support lookup-join, and the capabilities are falsely claiming that we have that capability, so we cannot use that here, but must disable lookup-join in CCS IT

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java [hunk 2]
```diff
@@ -206,23 +227,17 @@ protected static void checkCapabilities(RestClient client, TestFeatureService te
...
-            assumeTrue("Test " + testName + " requires " + feature, testFeatureService.clusterHasFeature(esqlFeature));
-        }
+        return false;
     }
 
     protected boolean supportsInferenceTestService() {
         return true;
     }
 
+    protected boolean supportsIndexModeLookup() throws IOException {
```
Comment (astefan) [inline x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:237]: I guess you didn't mean on having this method to `throws IOException`?

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java [hunk 3]
```diff
@@ -182,12 +182,33 @@ protected void shouldSkipTest(String testName) throws IOException {
...
         throws IOException {
-        if (testCase.requiredCapabilities.isEmpty()) {
+        if (hasCapabilities(client, testCase.requiredCapabilities)) {
             return;
         }
+
+        var features = Stream.concat(
+            new EsqlFeatures().getFeatures().stream(),
+            new EsqlFeatures().getHistoricalFeatures().keySet().stream()
+        ).map(NodeFeature::id).collect(Collectors.toSet());
```
Comment (astefan) [inline x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:192]: Since you moved this to a better place, do you think it's valid for this Set to be built only once per IT test class instance? Looking at `checkCapabilities` usage I _think_ this Set is called and built for every test query, which seems overkill.

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java (hunk 2 above)
Comment (astefan) [inline x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:237]: Imo, support for these special situations (inference and lookup index mode... so far) are a bit messy and we could improve this (in the future when maybe someone will get frustrated enough when managing an increasing list of exceptions) in the form of a `record` that has all these `boolean`s inside and instead of dealing with separate `boolean`s we deal with a single, grouped set of settings.

File: x-pack/plugin/esql/qa/server/mixed-cluster/build.gradle [hunk 4]
```diff
@@ -22,6 +22,7 @@ restResources {
 dependencies {
   javaRestTestImplementation project(xpackModule('esql:qa:testFixtures'))
   javaRestTestImplementation project(xpackModule('esql:qa:server'))
+  javaRestTestImplementation project(xpackModule('esql'))
```
Comment (astefan) [inline x-pack/plugin/esql/qa/server/mixed-cluster/build.gradle:25]: I don't see an obvious reason for this addition. Why did you need to add it?

**Review by astefan — APPROVED** [review]
LGTM, but left some comments.

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java (hunk 2 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:237]: Not originally, but the capabilities check in the sub-classes can throw the exception, so the parent signature needed it too.

File: x-pack/plugin/esql/qa/server/mixed-cluster/build.gradle (hunk 4 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/qa/server/mixed-cluster/build.gradle:25]: So the capabilities check can use the enum `JOIN_LOOKUP_V4` instead of a plan string, which I think would make it very likely if we changed the enum that we would get tests disabled that we did not want disabled. I saw Nhat added this to the MultiClusterSpecIT, so I did the same here.

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:192]: This code is called exactly as many times as before, from shouldSkipTest which is run for each test. We could move this collection to a class variable, but it seems a refinement out of scope for this PR.
